File: web/src/services/debrid/AllDebridService.ts
// Port of Sources/DebridStreamer/Services/Debrid/AllDebridService.swift.
//
// fetch-based AllDebrid client. Mirrors the Swift actor: chunked /magnet/instant
// cache check, /magnet/upload + /magnet/status poll, /link/unlock unrestrict, and
// /user account decoding. Auth is sent as a Bearer + X-API-Key header and an
// `apikey=` form-body component (never in the query). The poll SLEEP is
// injectable (default no-op) so tests don't wait.

import {
  CacheStatus,
  type DebridAccountInfo,
  type DebridFileCandidate,
  DebridFileSelector,
  type EpisodeFileHint,
  type DebridServiceType,
  DebridServiceType as DebridServiceTypeNS,
  type DebridTorrent,
  type StreamInfo,
  AudioFormat,
  SourceType,
  VideoCodec,
  VideoQuality,
} from "./models";
import {
  type DebridService,
  DebridError,
  type FetchImpl,
  defaultFetchImpl,
  formValueEncode,
} from "./types";
import {
  asObject,
  asObjectArray,
  int64Value,
  noopSleep,
  parseJSONObject,
  type Sleep,
} from "./jsonHelpers";

/** Chunk an array into sub-arrays of `size`. Mirrors the Swift `chunked(into:)`. */
function chunked<T>(arr: T[], size: number): T[][] {
  const out: T[][] = [];
  for (let i = 0; i < arr.length; i += size) out.push(arr.slice(i, i + size));
  return out;
}

export class AllDebridService implements DebridService {
  readonly serviceType: DebridServiceType = DebridServiceTypeNS.allDebrid;
  private readonly apiToken: string;
  private readonly baseURL = "https://api.alldebrid.com/v4";
  private readonly fetchImpl: FetchImpl;
  private readonly sleep: Sleep;
  private readonly agent = "DebridStreamer";

  constructor(apiToken: string, fetchImpl?: FetchImpl, sleep: Sleep = noopSleep) {
    this.apiToken = apiToken;
    this.fetchImpl = fetchImpl ?? defaultFetchImpl();
    this.sleep = sleep;
  }

  async checkCache(hashes: string[]): Promise<Record<string, CacheStatus>> {
    if (hashes.length === 0) return {};

    const results: Record<string, CacheStatus> = {};
    for (const chunk of chunked(hashes, 100)) {
      const magnetsParam = chunk.map((h) => `magnets[]=${h}`).join("&");
      const data = await this.requestRaw("/magnet/instant", "GET", magnetsParam);

      const json = parseJSONObject(data);
      const dataObj = json && asObject(json.data);
      const magnets = dataObj && asObjectArray(dataObj.magnets);
      if (magnets) {
        for (const magnetInfo of magnets) {
          // Skip rows without a real hash (error/invalid magnet entries have no
          // `hash` field): writing results[""] would pollute the map and leave
          // the actually-requested hash unrecorded (→ silently "not cached").
          if (typeof magnetInfo.hash !== "string" || magnetInfo.hash.length === 0) {
            continue;
          }
          const hash = magnetInfo.hash.toLowerCase();
          const instant = magnetInfo.instant === true;
          results[hash] = instant
            ? CacheStatus.cached(null, null, null)
            : CacheStatus.notCached;
        }
      }
    }

    return results;
  }

  async addMagnet(hash: string): Promise<string> {
    const magnet = `magnet:?xt=urn:btih:${hash}`;
    const body = `magnets[]=${formValueEncode(magnet)}`;
    const data = await this.requestRaw("/magnet/upload", "POST", undefined, body);

    const json = parseJSONObject(data);
    const dataObj = json && asObject(json.data);
    const magnets = dataObj && asObjectArray(dataObj.magnets);
    const first = magnets && magnets[0];
    const id = first ? int64Value(first.id) : null;
    if (id == null) {
      throw DebridError.downloadFailed("Failed to add magnet to AllDebrid");
    }
    return String(id);
  }

  async selectFiles(_torrentId: string, _fileIds: number[]): Promise<void> {
    // AllDebrid auto-selects files; this is a no-op for most cases.
  }

  async getStreamURL(
    torrentId: string,
    fileHint: EpisodeFileHint | null = null,
  ): Promise<StreamInfo> {
    const maxAttempts = 20;
    let magnets: Record<string, unknown> = {};
    for (let attempt = 0; attempt < maxAttempts; attempt++) {
      const data = await this.requestRaw(
        "/magnet/status",
        "GET",
        `id=${torrentId}`,
      );
      const json = parseJSONObject(data);
      const dataObj = json && asObject(json.data);
      const magnet = dataObj && asObject(dataObj.magnets);
      if (magnet == null) throw DebridError.torrentNotFound(torrentId);
      magnets = magnet;
      const status = typeof magnet.status === "string" ? magnet.status : "";
      if (status === "Ready") break;
      if (status === "Error" || status.toLowerCase().includes("error")) {
        throw DebridError.downloadFailed(`AllDebrid reported status: ${status}`);
      }
      if (attempt === maxAttempts - 1) {
        throw DebridError.downloadFailed(
          `Torrent not ready after ${maxAttempts}s (status: ${status})`,
        );
      }
      await this.sleep(1000);
    }

    const links = asObjectArray(magnets.links);
    if (links == null) throw DebridError.noFilesAvailable();

    const candidates: DebridFileCandidate[] = [];
    for (const item of links) {
      if (typeof item.link !== "string") continue;
      const filename = typeof item.filename === "string" ? item.filename : "Unknown";
      const size = int64Value(item.size) ?? 0;
      candidates.push({ link: item.link, fileName: filename, sizeBytes: size });
    }
    const selected = DebridFileSelector.selectBest(candidates, fileHint);
    if (selected == null) throw DebridError.noFilesAvailable();

    const streamURL = await this.unrestrict(selected.link);
    const filename = selected.fileName;
    const size = selected.sizeBytes;

    return {
      streamURL,
      quality: VideoQuality.parse(filename),
      codec: VideoCodec.parse(filename),
      audio: AudioFormat.parse(filename),
      source: SourceType.parse(filename),
      sizeBytes: size,
      fileName: filename,
      debridService: "AD",
    };
  }

  async unrestrict(link: string): Promise<string> {
    const body = `link=${formValueEncode(link)}`;
    const data = await this.requestRaw("/link/unlock", "POST", undefined, body);

    const json = parseJSONObject(data);
    const dataObj = json && asObject(json.data);
    const downloadStr = dataObj && typeof dataObj.link === "string" ? dataObj.link : null;
    if (downloadStr == null || !isAbsoluteURL(downloadStr)) {
      throw DebridError.downloadFailed("Failed to unrestrict link on AllDebrid");
    }
    return downloadStr;
  }

  /** List the account's magnets (the Debrid Library manager source). Calls
   * `GET /magnet/status` with no id, which returns every magnet under
   * `data.magnets`, normalizing each into a `DebridTorrent`. Fault-tolerant:
   * an unparseable body yields an empty list. */
  async listTorrents(): Promise<DebridTorrent[]> {
    const data = await this.requestRaw("/magnet/status", "GET");
    const json = parseJSONObject(data);
    const dataObj = json && asObject(json.data);
    const magnets = dataObj && asObjectArray(dataObj.magnets);
    if (magnets == null) return [];
    return magnets.map((m) => normalizeADTorrent(m));
  }

  /** Delete a magnet from the account by id (`GET /magnet/delete?id=`). */
  async deleteTorrent(id: string): Promise<void> {
    await this.requestRaw("/magnet/delete", "GET", `id=${id}`);
  }

  async validateToken(): Promise<boolean> {
    try {
      await this.getAccountInfo();
      return true;
    } catch {
      return false;
    }
  }

  async getAccountInfo(): Promise<DebridAccountInfo> {
    const data = await this.requestRaw("/user", "GET");
    const json = parseJSONObject(data);
    const dataObj = json && asObject(json.data);
    const user = dataObj && asObject(dataObj.user);
    if (user == null) throw DebridError.invalidToken();

    const username = typeof user.username === "string" ? user.username : "Unknown";
    const email = typeof user.email === "string" ? user.email : null;
    const isPremium = user.isPremium === true;

    let premiumExpiry: Date | null = null;
    if (typeof user.premiumUntil === "number") {
      premiumExpiry = new Date(user.premiumUntil * 1000);
    }

    return { username, email, premiumExpiry, isPremium };
  }

  // MARK: - HTTP

  private async requestRaw(
    path: string,
    method: string,
    queryParams?: string,
    body?: string,
  ): Promise<string> {
    let urlStr = `${this.baseURL}${path}?agent=${this.agent}`;
    if (queryParams != null && queryParams.length > 0) {
      urlStr += `&${queryParams}`;
    }

    const headers: Record<string, string> = {
      Authorization: `Bearer ${this.apiToken}`,
      "X-API-Key": this.apiToken,
    };

    let composedBody: string | undefined;
    if (body != null) {
      const authBodyComponent = `apikey=${formValueEncode(this.apiToken)}`;
      composedBody = body.length === 0 ? authBodyComponent : `${body}&${authBodyComponent}`;
      headers["Content-Type"] = "application/x-www-form-urlencoded";
    }

    const response = await this.fetchImpl(urlStr, {
      method,
      headers,
      body: composedBody,
    });
    const status = response.status;

    if (!(status >= 200 && status <= 299)) {
      if (status === 401) throw DebridError.invalidToken();
      const errText = (await response.text().catch(() => "")) || "";
      throw DebridError.httpError(status, errText);
    }

    return response.text();
  }
}

function isAbsoluteURL(value: string): boolean {
  try {
    new URL(value);
    return true;
  } catch {
    return false;
  }
}

/** Normalize a raw AllDebrid `/magnet/status` magnet into the shared
 * `DebridTorrent` display shape. Defensive against missing/typed-wrong fields. */
function normalizeADTorrent(m: Record<string, unknown>): DebridTorrent {
  const idNum = int64Value(m.id);
  const id = idNum != null ? String(idNum) : String(m.id ?? "");
  const name = typeof m.filename === "string" ? m.filename : "Unknown";
  const hash = typeof m.hash === "string" ? m.hash.toLowerCase() : null;
  const status = typeof m.status === "string" ? m.status : "unknown";
  const bytes = int64Value(m.size) ?? 0;
  // AllDebrid reports `uploadDate` as a unix epoch (seconds).
  let addedAt: string | null = null;
  const upload = int64Value(m.uploadDate);
  if (upload != null && upload > 0) addedAt = new Date(upload * 1000).toISOString();
  return {
    id,
    name,
    sizeBytes: bytes,
    status,
    infoHash: hash,
    addedAt,
    host: null,
    progress: null,
    debridService: "AD",
  };
}

File: Sources/DebridStreamer/Services/Debrid/AllDebridService.swift
import Foundation

/// AllDebrid API integration.
/// API docs: https://docs.alldebrid.com/
actor AllDebridService: DebridServiceProtocol {
    let serviceType = DebridServiceType.allDebrid
    private let apiToken: String
    private let baseURL = "https://api.alldebrid.com/v4"
    private let session: URLSession
    private let agent = "DebridStreamer"

    init(apiToken: String, session: URLSession = AppHTTP.api) {
        self.apiToken = apiToken
        self.session = session
    }

    func checkCache(hashes: [String]) async throws -> [String: CacheStatus] {
        guard !hashes.isEmpty else { return [:] }

        var results: [String: CacheStatus] = [:]
        let chunks = hashes.chunked(into: 100)

        for chunk in chunks {
            let magnetsParam = chunk.map { "magnets[]=\($0)" }.joined(separator: "&")
            let data = try await requestRaw(
                path: "/magnet/instant",
                method: "GET",
                queryParams: magnetsParam
            )

            if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
               let dataObj = json["data"] as? [String: Any],
               let magnets = dataObj["magnets"] as? [[String: Any]] {
                for magnetInfo in magnets {
                    let hash = (magnetInfo["hash"] as? String)?.lowercased() ?? ""
                    let instant = magnetInfo["instant"] as? Bool ?? false
                    if instant {
                        results[hash] = .cached(fileId: nil, fileName: nil, fileSize: nil)
                    } else {
                        results[hash] = .notCached
                    }
                }
            }
        }

        return results
    }

    func addMagnet(hash: String) async throws -> String {
        let magnet = "magnet:?xt=urn:btih:\(hash)"
        let body = "magnets[]=\(magnet.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? magnet)"
        let data = try await requestRaw(path: "/magnet/upload", method: "POST", body: body)

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let dataObj = json["data"] as? [String: Any],
              let magnets = dataObj["magnets"] as? [[String: Any]],
              let first = magnets.first,
              let id = first["id"] as? Int else {
            throw DebridError.downloadFailed("Failed to add magnet to AllDebrid")
        }
        return String(id)
    }

    func selectFiles(torrentId: String, fileIds: [Int]) async throws {
        // AllDebrid auto-selects files; this is a no-op for most cases
    }

    func getStreamURL(torrentId: String) async throws -> StreamInfo {
        // Poll the magnet status until it becomes "Ready" (mirrors the bounded
        // retry loops used by Real-Debrid and TorBox). A freshly-added magnet is
        // rarely "Ready" on the first request, so a single read would fail for any
        // non-cached torrent.
        let maxAttempts = 20
        var magnets: [String: Any] = [:]
        for attempt in 0..<maxAttempts {
            let data = try await requestRaw(path: "/magnet/status", method: "GET", queryParams: "id=\(torrentId)")
            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let dataObj = json["data"] as? [String: Any],
                  let magnet = dataObj["magnets"] as? [String: Any] else {
                throw DebridError.torrentNotFound(torrentId)
            }
            magnets = magnet
            let status = (magnet["status"] as? String) ?? ""
            if status == "Ready" { break }
            // Terminal failure states — stop polling early.
            if status == "Error" || status.lowercased().contains("error") {
                throw DebridError.downloadFailed("AllDebrid reported status: \(status)")
            }
            if attempt == maxAttempts - 1 {
                throw DebridError.downloadFailed("Torrent not ready after \(maxAttempts)s (status: \(status))")
            }
            try await Task.sleep(nanoseconds: 1_000_000_000)
        }

        guard let links = magnets["links"] as? [[String: Any]] else {
            throw DebridError.noFilesAvailable
        }

        let candidates = links.compactMap { item -> DebridFileCandidate? in
            guard let link = item["link"] as? String else { return nil }
            let filename = item["filename"] as? String ?? "Unknown"
            let size = int64Value(item["size"]) ?? 0
            return DebridFileCandidate(link: link, fileName: filename, sizeBytes: size)
        }
        guard let selected = DebridFileSelector.selectBest(from: candidates) else {
            throw DebridError.noFilesAvailable
        }

        let streamURL = try await unrestrict(link: selected.link)
        let filename = selected.fileName
        let size = selected.sizeBytes

        return StreamInfo(
            streamURL: streamURL.absoluteString,
            quality: VideoQuality.parse(from: filename),
            codec: VideoCodec.parse(from: filename),
            audio: AudioFormat.parse(from: filename),
            source: SourceType.parse(from: filename),
            sizeBytes: size,
            fileName: filename,
            debridService: "AD"
        )
    }

    func unrestrict(link: String) async throws -> URL {
        let body = "link=\(link.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? link)"
        let data = try await requestRaw(path: "/link/unlock", method: "POST", body: body)

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let dataObj = json["data"] as? [String: Any],
              let downloadStr = dataObj["link"] as? String,
              let url = URL(string: downloadStr) else {
            throw DebridError.downloadFailed("Failed to unrestrict link on AllDebrid")
        }
        return url
    }

    func validateToken() async throws -> Bool {
        do {
            _ = try await getAccountInfo()
            return true
        } catch {
            return false
        }
    }

    func getAccountInfo() async throws -> DebridAccountInfo {
        let data = try await requestRaw(path: "/user", method: "GET")

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let dataObj = json["data"] as? [String: Any],
              let user = dataObj["user"] as? [String: Any] else {
            throw DebridError.invalidToken
        }

        let username = user["username"] as? String ?? "Unknown"
        let email = user["email"] as? String
        let isPremium = user["isPremium"] as? Bool ?? false

        var premiumExpiry: Date?
        if let expiryTimestamp = user["premiumUntil"] as? TimeInterval {
            premiumExpiry = Date(timeIntervalSince1970: expiryTimestamp)
        }

        return DebridAccountInfo(
            username: username,
            email: email,
            premiumExpiry: premiumExpiry,
            isPremium: isPremium
        )
    }

    // MARK: - HTTP

    private func requestRaw(path: String, method: String, queryParams: String? = nil, body: String? = nil) async throws -> Data {
        var urlStr = baseURL + path + "?agent=\(agent)"
        if let queryParams = queryParams, !queryParams.isEmpty {
            urlStr += "&\(queryParams)"
        }

        guard let url = URL(string: urlStr) else {
            throw DebridError.networkError("Invalid URL: \(path)")
        }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("Bearer \(apiToken)", forHTTPHeaderField: "Authorization")
        request.setValue(apiToken, forHTTPHeaderField: "X-API-Key")

        if let body = body {
            let authBodyComponent = "apikey=\(apiToken.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? apiToken)"
            let composedBody = body.isEmpty ? authBodyComponent : "\(body)&\(authBodyComponent)"
            request.httpBody = composedBody.data(using: .utf8)
            request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        }

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw DebridError.networkError("Invalid response")
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            if httpResponse.statusCode == 401 {
                throw DebridError.invalidToken
            }
            throw DebridError.httpError(httpResponse.statusCode, String(data: data, encoding: .utf8) ?? "")
        }

        return data
    }

    private func int64Value(_ value: Any?) -> Int64? {
        if let number = value as? NSNumber {
            return number.int64Value
        }
        if let intValue = value as? Int {
            return Int64(intValue)
        }
        if let stringValue = value as? String {
            return Int64(stringValue)
        }
        return nil
    }
}
